src/cursor/abstract_cursor.ts

```diff
@@ -0,0 +1,710 @@
+    callback?: Callback<void>
+  ): Promise<void> | void {
+    if (typeof iterator !== 'function') {
+      throw new TypeError('Missing required parameter `iterator`');
+    }
+
+    return maybePromise(callback, done => {
+      const transform = this[kTransform];
+      const fetchDocs = () => {
+        next(this, (err, doc) => {
+          if (err || doc == null) return done(err);
+          if (doc == null) return done();
+
+          // NOTE: no need to transform because `next` will do this automatically
+          let result = iterator(doc);
+          if (result === false) return done();
+
+          // these do need to be transformed since they are copying the rest of the batch
+          const internalDocs = this[kDocuments].splice(0, this[kDocuments].length);
+          if (internalDocs) {
+            for (let i = 0; i < internalDocs.length; ++i) {
+              result = iterator(transform ? transform(internalDocs[i]) : internalDocs[i]);
+              if (result === false) return done();
+            }
+          }
+
+          fetchDocs();
+        });
+      };
+
+      fetchDocs();
+    });
+  }
+
+  close(): void;
+  close(callback: Callback): void;
+  close(options: CursorCloseOptions): Promise<void>;
+  close(options: CursorCloseOptions, callback: Callback): void;
+  close(options?: CursorCloseOptions | Callback, callback?: Callback): Promise<void> | void {
+    if (typeof options === 'function') (callback = options), (options = {});
+    options = options || {};
+
+    const needsToEmitClosed = !this[kClosed];
+    this[kClosed] = true;
+
+    return maybePromise(callback, done => {
+      const cursorId = this[kId];
+      const cursorNs = this[kNamespace];
+      const server = this[kServer];
+      const session = this[kSession];
+
+      if (cursorId == null || server == null || cursorId.isZero() || cursorNs == null) {
+        if (needsToEmitClosed) {
+          this[kId] = Long.ZERO;
+          this.emit(AbstractCursor.CLOSE);
+        }
+
+        if (session && session.owner === this) {
+          return session.endSession(done);
+        }
+
+        return done();
+      }
+
+      this[kKilled] = true;
+      server.killCursors(
+        cursorNs.toString(),
+        [cursorId],
+        { ...pluckBSONSerializeOptions(this[kOptions]), session },
+        () => {
+          if (session && session.owner === this) {
+            return session.endSession(() => {
+              this.emit(AbstractCursor.CLOSE);
+              done();
+            });
+          }
+
+          this.emit(AbstractCursor.CLOSE);
+          done();
+        }
+      );
+    });
+  }
+
+  /**
+   * Returns an array of documents. The caller is responsible for making sure that there
+   * is enough memory to store the results. Note that the array only contains partial
+   * results when this cursor had been previously accessed. In that case,
+   * cursor.rewind() can be used to reset the cursor.
```

Comment:
Should remove the reference to `cursor.rewind()` or add a note about NODE-2811 here.